packages/pwa-kit-react-sdk/src/ssr/universal/components/refresh/index.jsx

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright (c) 2023, Salesforce, Inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+import React, {useEffect} from 'react'
+import {useHistory, useLocation} from 'react-router-dom'
+import {useQueryClient} from '@tanstack/react-query'
+
+// For good UX, show loading spinner long enough for users to see
+const LOADING_SPINNER_MIN_DURATION = 500
+
+/**
+ * A _private_ component to show loading spinner while refetching data on the client-side.
+ * To trigger this refetch, we do soft navigation back to the referrer.
+ * So this component is meant to be used as a route with `referrer` search param.
+ *
+ * @example
+ * const navigate = useNavigation()
+ * navigate(`/__refetch-data?referrer=${encodeURIComponent(urlOfCurrentPage)}`, 'replace')
+ *
+ * @private
+ */
+const Refresh = () => {
+    const history = useHistory()
+    const location = useLocation()
+
+    let queryClient
+    try {
+        queryClient = useQueryClient()
```

Comment:
Even though this PR is for V2 SDK...  if I understand it correctly, V2.3 starts supporting react-query use. So we'll need to consider the possibility of react-query even here (and not only in V3).